lib/bandit/headers.ex

```diff
@@ -0,0 +1,65 @@
+defmodule Bandit.Headers do
+  @moduledoc false
+  # Conveniences for dealing with headers
+
+  # credo:disable-for-this-file Credo.Check.Design.AliasUsage
+
+  def get_header(headers, header) do
+    case List.keyfind(headers, header, 0) do
+      {_, value} -> value
+      nil -> nil
+    end
+  end
+
+  def parse_hostlike_header(host_header) do
+    host_header
+    |> :binary.split(":")
+    |> case do
+      [host, port] ->
+        case Integer.parse(port) do
+          {port, ""} when port > 0 -> {:ok, host, port}
+          _ -> {:error, "Header contains invalid port"}
+        end
+
+      [host] ->
+        {:ok, host, nil}
+    end
+  end
+
+  def get_content_length(headers) do
+    with value when not is_nil(value) <- get_header(headers, "content-length"),
+         {:ok, length} <- parse_content_length(value) do
+      if length >= 0 do
+        {:ok, length}
+      else
+        {:error, "invalid negative content-length header (RFC9110§8.6)"}
+      end
+    else
+      nil -> {:ok, nil}
+      error -> error
+    end
+  end
+
+  defp parse_content_length(value) do
+    case Integer.parse(value) do
+      {length, ""} ->
+        {:ok, length}
+
+      {length, rest} ->
+        rest
+        |> Plug.Conn.Utils.list()
+        |> enforce_unique_value(to_string(length), length)
+
+      :error ->
+        {:error, "invalid content-length header (RFC9112§6.3.5)"}
+    end
+  end
```

Comment:
If you put both the nil case handling and the length check and error case in the parse function then both error cases are there and the with can be simplified.

```suggestion
  def get_content_length(headers) do
    with value <- get_header(headers, "content-length") do
      parse_content_length(value)
    end
  end

  defp parse_content_length(nil), do: {:ok, nil}
  
  defp parse_content_length(value) do
    case Integer.parse(value) do
      {length, ""} when length >= 0 ->
        {:ok, length}
        
      {length, ""} ->
        {:error, "invalid negative content-length header (RFC9110§8.6)"}

      {length, rest} ->
        rest
        |> Plug.Conn.Utils.list()
        |> enforce_unique_value(to_string(length), length)

      :error ->
        {:error, "invalid content-length header (RFC9112§6.3.5)"}
    end
  end
  ```

Reply:
or should I hold my tongue until it's not a draft? 😅

Reply:
Alternatively, if you don't  feel good about a `nil` case for `parse_content_length`

```suggestion
  def get_content_length(headers) do
    case get_header(headers, "content-length") do
      nil -> {:ok, nil}
      value -> parse_content_length(value)
    end
  end

  defp parse_content_length(value) do
    case Integer.parse(value) do
      {length, ""} when length >= 0 ->
        {:ok, length}
        
      {length, ""} ->
        {:error, "invalid negative content-length header (RFC9110§8.6)"}

      {length, rest} ->
        rest
        |> Plug.Conn.Utils.list()
        |> enforce_unique_value(to_string(length), length)

      :error ->
        {:error, "invalid content-length header (RFC9112§6.3.5)"}
    end
  end
  ```

Reply:
This is rapidly becoming 'not a draft' (I have a few more changes but what's up here won't substantially change I don't expect). 

Paging @moogle19 to have an opinion on the content length changes as this is a literal copy and paste of code they added a few weeks ago. Looks fine to me, I think. 

The only thing to consider is that this is one place that speed should win over clarity; it's an incredibly hot path (not that I think that there's anything slow in your changes, just that the guiding light here is a bit different)

Reply:
Yeah, I don't expect to see any difference in performance (unless removing the guard clause for `nil` actually helps), but it removes an extra nested conditional and would keep the content-length RFC errors in the parsing function instead of spreading them around.

Reply:
Yeah it's a good change

Reply:
I'm not smart enough to contribute features, but I can nitpick the heck out of some code 😅 😬

Reply:
Went with your second suggestion; it reads great. 

Thanks again!